### DISCUSSION TITLE
Assistant response post-processing

### DISCUSSION THREAD

**Discussion by cpursley** [brainlid/langchain#130] [Ideas]
Ability to automatically run additional post-processing on the last `:assistant` response.

###  Examples:

- Run a data extraction on the response (my use case is generating list of followup questions)
- Pipe contents of last message following a successful tool results call into a better/different chat model
- Run a quality/fact/formatting check on the contents of the last assistant response

**Comment by cpursley** [comment]
💻  It appears there's some related work on this branch: https://github.com/brainlid/langchain/tree/me-destructuring

So maybe this could be expanded to optionally take in different llm models, as some are better suited for certain tasks than others.

**Reply by brainlid** [reply]
Correct! I will soon publish an RC for the new release that adds support for a plug-like middleware that processes assistant response messages. The first built-in "message processor" is `JsonProcessor`. It supports extracting a JSON response and converting it to an Elixir map. Because LLMs vary in their ability to support JSON reliably, it automatically returns errors to the LLM to be fixed. A processor is best when coupled with a new `|> LLMChain.run(mode: :until_success)` that repeats the round-trip requests for the LLM to fix something. The `LLMChain` has a new `max_retry_count` that is prevents it from cycling over and over until the token limit is exhausted.

**Reply by brainlid** [reply]
You can create custom processors as well. Like plug, they are applied in order. So you could add a fact/formatting check for your specific needs. It is just a function so it's very easy to create your own.

**Reply by cpursley** [reply]
@brainlid I'm testing out the new changes (thank you btw). Now I get this sort of output after a new LLMChain.run(mode: :until_success) that has functions:

```elixir
%LangChain.Message{
  content: nil,
  processed_content: nil,
  index: nil,
  status: :complete,
  role: :tool,
  name: nil,
  tool_calls: [],
  tool_results: [
    %LangChain.Message.ToolResult{
      type: :function,
      tool_call_id: "call_j7YIpXgdad2PZkFb1iLQjC6P",
      name: "get_details",
      content: "some details json",
      display_text: nil,
      is_error: false
    },
    %LangChain.Message.ToolResult{
      type: :function,
      tool_call_id: "call_apiFKJzzspYisF4TVOplzh2m",
      name: "calculator",
      content: "8",
      display_text: nil,
      is_error: false
    }
  ]
}
```

The idea is we now create a processor handler to iterate through these tool results? If so, what might that look like?

**Reply by brainlid** [reply]
@cpursley no, a Message Processor handles assistant responses. It would be used in a situation where you are not using a tool to get the final data you want. It works better for a situation where you instruct it to give a final result in JSON like this:

```
Output in the following JSON format:

{
  "alt": "generated alt text",
  "caption": "generation caption text"
}
```

Then a processor would convert the JSON to an Elixir map.

I've been thinking about cleaner ways to use a ToolResult. I'm not happy with it at the moment. It's basically a pattern match or imperative code to pull the data out.

In your example, you may want to use `mode: :while_needs_response` to have the results of the 2 tool calls returned to the LLM so it can combine then into an answer. You could instruct it to give you the final answer in JSON format if that works best. For simple JSON, you can define it in the prompt. For more complex JSON structures, you can use the `LangChain.FunctionParam` to build up the  structure, then provide it to the LLM in JSONSchema format using `to_parameters_schema` or `to_json_schema` depending on what best fits your situation.

I'm just basing that on the fact that 2 different tools are being used.

**Reply by brainlid** [reply]
If the result you wanted IS the tool call with the argument in a JSON form, then in your tool function, you can send the response to your LiveView (if that makes sense). A bit more detail on your use-case would be helpful.

**Reply by cpursley** [reply]
Basically what I'd like to do is send the result of the tool calls (like above) to a different llm model that's better than the tool calling one. Maybe I could iterate over these tools results and send those over a separate chain to the other llm model, then put it's results into a completed message struct and then add back to the main chain? Or could I do that via a custom processor?

**Reply by cpursley** [reply]
Update: custom processor seems to accomplish what I'm trying to do. I'm creating a separate chain in the process function and returning the content from that.

Which makes me think a "post-processor chain" thing that could take it's own instructions and different llm and match on certain conditions of tool results or content of single message response following tool results would be nice (if that makes sense). For doing things like cleaning up responses, reformatting, quality checking, etc. 

It also might make sense to somehow tag the message following a tool results (maybe set a name).

**Reply by brainlid** [reply]
The message processors currently aren't called to evaluate and operate on a ToolResult. However, another option that could work comes from something I saw in the JS LangChain project with the callbacks. There's the idea of an event notification like an "on_llm_new_delta" event that just notifies you it was created, then there was a "handle_" style event. Perhaps that could be a kind of processors where you could perform some operation/formatting/etc and the return replaces the message. Like a `handle_tool_response" or something.

But maybe a custom processors that DOES get called for a Tool response is a better move. Keeping things simpler and all together more.

**Reply by brainlid** [reply]
What is the ToolCall doing? Are you using it as a way for the LLM to extract structured data? Or is it actually doing something in your application like executing an Ecto query and returning the result? It sounds almost like you're using it as a more detailed router for extracting data but using a function to do it. Is that right?

**Reply by cpursley** [reply]
The tool call is currently making an http call to 3rd party services. But I'be got plans for tools that ecto queries as well (like the fitness app). Main thing I'm after is the tool calling model I'm using isn't great at general purpose llm things and I want to process the tool call results with the better llm.

**Reply by cpursley** [reply]
Speaking of routing, it would also be awesome (if feasible) to have a smart prompt router that worked per prompt. For example, do a quick analysis on the user prompt and then decide which different model to use (if any). But all on the same chain instead of setting up a new chain. Not sure if that makes sense. Mainly, some models are better at certain tasks and being able to dynamically switch between them would be useful.

**Reply by brainlid** [reply]
@cpursley in my experience, a smaller/faster/lesser model doesn't do as well with tool calling. I think it will work better to have a single simple router chain examine the user's prompt and decide where it should go from there. Just route it to a chain setup for the purpose. 

The selected chain can have the higher quality model you want and the appropriate tools loaded for it's purpose. You re-run the user's original prompt in the appropriate chain and let it do the ToolCalls instead of having that done on a less capable model.

**Reply by cpursley** [reply]
Funny, I came to a similar conclusion but just without the router. Just run one chain first that can do function calling and feed those results (I used :until_success and mapped out the ToolResult content into the stronger llm's prompt). Works pretty well.

The struggle I'm having is llama tool calling model I'm using kindof sucks for everything else (it's a one-trick pony). But giving those tool call json results to regular llama instruct model works well.

**Reply by brainlid** [reply]
@cpursley 

> it would also be awesome (if feasible) to have a smart prompt router that worked per prompt. For example, do a quick analysis on the user prompt and then decide which different model to use (if any). But all on the same chain instead of setting up a new chain.

I think you want an Agent.

Many smaller self-hosted models do 1 thing well 1 time. Meaning I can ask it to look at the user's prompt and have it decide if the user wants setup_1, setup_2 or setup_3. If you try to ask a follow-up question on that same model/chain, it will hallucinate to no end. It can do the 1 job 1 time. Any future uses need to start over.

I think what you really want is an Agent. The Agent internally manages multiple chains for multiple purposes but you only talk to the one Agent.

**Comment by brainlid** [comment]
Yes, with v0.3.0-rc.0 (what's currently on main), a new feature called MessageProcessors was added. A JsonProcessor is provided but custom ones can be created. The purpose is to process an Assistant response message to parse, change formatting, enforce rules, etc. When the assistant's response doesn't meet a rule, the processor generates a user message reporting the error to let the LLM try again.
